ddl/job_table.go

```diff
@@ -194,28 +165,41 @@ func hasSysDB(job *model.Job) bool {
 	return false
 }
 
-func (d *ddl) handleUpgradingState(se *sess.Session, job *model.Job) error {
-	if !d.stateSyncer.IsUpgradingState() {
-		if !job.IsPausedBySystem() || hasSysDB(job) {
-			return nil
+func (d *ddl) processJobDuringUpgrade(sess *sess.Session, job *model.Job) (err error, isRunnable bool) {
+	if d.stateSyncer.IsUpgradingState() {
+		if hasSysDB(job) {
+			return nil, true
 		}
-		_, err := ResumeJobsBySystem(se.Session(), []int64{job.ID})
-		if err != nil {
-			logutil.BgLogger().Warn("[ddl] resume user DDL by system failed", zap.Stringer("job", job), zap.Error(err))
-			return err
+		var errs []error
+		// During binary upgrade, pause all running DDL jobs
+		errs, err = PauseJobsBySystem(sess.Session(), []int64{job.ID})
+		if len(errs) > 0 {
+			if dbterror.ErrCannotPauseDDLJob.Equal(errs[0]) ||
+				dbterror.ErrPausedDDLJob.Equal(errs[0]) {
+				return nil, false
+			}
+
+			// During upgrade, there are jobs that may not be paused because of `!job.IsRunning`.
+			// Then, we would not run it and just return unless the upgrade finished.
+			return errs[0], false
 		}
-		logutil.BgLogger().Info("[ddl] resume user DDL by system successful", zap.Stringer("job", job))
-		return nil
-	}
-	if job.IsPausing() {
-		return nil
+
+		// no matter how, we would not run DDL jobs during upgrade even not paused.
+		return err, false
```

Comment:
If the job is pausing, we can't do `pauseReorgWorkers`, so I think we need to add the check of `!job.IsPausedBySystem() ` in line170.

Reply:
pausing is kind of request, it will be turned to be 'model.JobStatePaused' in pauseReorgWorkers() in ddl worker.

Reply:
We may let it keep running if the error is ErrPausedDDLJob.